README.md

This 'calorie' counter is based on a Cloudflare Durable Object, so we don't need to have a full database set up for our counter.

Find out more about these in the official [Durable Object documentation](https://developers.cloudflare.com/workers/learning/using-durable-objects).

Worker code is in `src/`. The Durable Object `Counter` class is in `src/counter.mjs`, and the frontend script is in `index.mjs`.


# API

There are currently just three REST API endpoints: `/reset`, `/increment` and read the counter via `/`

Some of the priviledge APIs require a "server" secret be passed via a `secret=foobar` query parameter.

The secret is set via `wrangler secret put SERVER_SECRET` and can be shared with whoever needs to use these APIs

## /reset

The reset API is simply invoked like `GET http://host/reset?secret=12345`

## /increment
The increment API is invoked like `GET http://host/increment?by=500&expiry=12345678&mac=0f94813426048e4301cd4fd1f3c7d6b86d6c6777452bcf7bbeb8a4fab8245634`

The `by` parameter indicates how much to increment the calorie count (which currently has a maximum of 2000 just to limit the damage of bugs with us incrementing the counter)

The `expiry` limits how long the request URL is valid for, to reduce the chance that people will discover the URL and then replay the request to manually increment the counter.

The `expiry` value should be set to a unix timestamp (in seconds), such as 60 seconds in the future. In python this would be `int(time.time()) + 60`

The `mac` parameter is a SHA-256 hash of an ascii/utf8 string formatted like this: `"/increment:by:expiry"`, e.g. `"/increment:500:12345678"`

The secret for the hash is set via `wrangler secret put APP_SECRET` and this secret is shared with each particpating app.

It's not an ideal setup, but hopefully ok for our current needs.

# /server_increment

This endpoint can be used to increment the counter from backends where there's no need to worry about leaking the secret.

Invoked like `GET http://host/increment?by=500&secret=1234`

The `by` parameter is the same as for `/increment`

# /daily_snapshots

Invoked like `GET http://host/daily_snapshots`

This will return a JSON result like:
```json
{
    "ver": 1,
    "snapshots": [
        [
            1643587200,
            500
        ],
    ]
}
```

with up to 30 daily snapshots.

Each snapshot is a pair like `[timestamp, counter]` where the timestamp is time since the unix epoch in seconds.

# /hourly_snapshots

Invoked like `GET http://host/hourly_snapshots`

This will return a JSON result like:
```json
{
    "ver": 1,
    "snapshots": [
        [
            1643648400,
            500
        ],
        [
            1643652000,
            200
        ]
    ]
}
```

with up to 24 hourly snapshots.

Each snapshot is a pair like `[timestamp, counter]` where the timestamp is time since the unix epoch in seconds.


# In-Game Integration

For anyone working with C# these utilities could be used to create signed `/increment` urls:

```csharp
using System.Text;
using System.Security.Cryptography;

private static string ToHexDigest(byte[] digest)
{
    StringBuilder hexDigest = new StringBuilder(digest.Length * 2);

    for (int i = 0; i < digest.Length; i++)
    {
        hexDigest.Append(digest[i].ToString("x2"));
    }

    return hexDigest.ToString();
}

private static string SignIncrement(byte[] key, int by, long expiry)
{
    var payload = $"/increment:{by}:{expiry}";
    using (HMACSHA256 hmac = new HMACSHA256(INDIE_GANG_API_KEY))
    {
        var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(payload));
        return ToHexDigest(hash);
    }
}

private const string INDIE_GANG_API_HOST = "indie-gang-api.realfit.co";
private static byte[] INDIE_GANG_API_KEY = Encoding.UTF8.GetBytes("asdf1234");

private static string CreateSignedIncrementUrl(int by)
{
    StringBuilder urlBuilder = new StringBuilder("http://", INDIE_GANG_API_HOST.Length + 128);
    long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
    long expiry = now + 60;

    var hexDigest = SignIncrement(INDIE_GANG_API_KEY, by, expiry);

    urlBuilder.Append(INDIE_GANG_API_HOST);
    urlBuilder.Append("/increment?by=");
    urlBuilder.Append(by);
    urlBuilder.Append("&expiry=");
    urlBuilder.Append(expiry);
    urlBuilder.Append("&mac=");
    urlBuilder.Append(hexDigest);

    return urlBuilder.ToString();
}
```

# Deploying

Deploying requires a Cloudflare account + domain and payment plan that supports durable objects (configured in `wrangler.toml`)

For now this is configured as a route under indie-gang-api.realfit.co. I can probably share an API key for others to deploy updates

Note: You must use [wrangler](https://developers.cloudflare.com/workers/cli-wrangler/install-update) 1.19.3 or newer to publish/deploy.


# Testing

There is a very simple test for the `/increment` endpoint in `test-signed-increment.py` which also shows how to create a signed url for submitting updates

To run this script first create a `config.json` like:

```json
{
    "increment_secret": "foobar",
    "host": "http://indie-gang-api.realfit.co"
}
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2022-02-06
### Added
- Added `/increment_server` API for incrementing via `secret` param, without needing to sign URLs
- Added daily/hourly snapshots and `/daily_snapshots` + `/hourly_snapshots` endpoints
- Renamed test tool to `test-api.py` and moved expected config to `../config.json` to keep
  it out of the repo and avoid risk of adding to source control
- Config should now look like:

```
{
    "app_secret": "foobar",
    "server_secret": "foobar",
    "host": "https://indie-gang-api.realfit.co"
}
```

test-api.py
#!/usr/bin/env python3

from http import server
import sys
import time
import hmac
import hashlib
import base64
import requests
import urllib
import json
import argparse

parser = argparse.ArgumentParser(description='Tool for smoke testing API')
parser.add_argument("--verbose", action="store_true", help="Verbose output")
default_config="../config.json"
parser.add_argument('--config', default=default_config, help="Config file including secrets (default = %s)" % default_config)

subparsers = parser.add_subparsers(help='sub-command help', dest='subcommand')

read_parser = subparsers.add_parser('read', help='Simply read counter')

inc_parser = subparsers.add_parser('increment', help='increment the calorie counter')
inc_parser.add_argument('by', type=int, help='How much to increment by')
inc_parser.add_argument("--server-api", action="store_true", help="Increment via server API instead of app/client API")

reset_parser = subparsers.add_parser('reset', help='reset the calorie counter')

daily_snaps_parser = subparsers.add_parser('daily_snapshots', help='Get the daily snapshots')

hourly_snaps_parser = subparsers.add_parser('hourly_snapshots', help='Get the hourly snapshots')

test_parser = subparsers.add_parser('test', help='Smoke test the API')

args = parser.parse_args()

config = None
with open(args.config) as fp:
    config = json.load(fp)
if config == None:
    sys.exit("Failed to load " + args.config)

host = config['host']
app_secret = config['app_secret'].encode('UTF-8')
server_secret = config['server_secret'].encode('UTF-8')


def create_signed_increment_url(by, expire_delta=60):
    global host
    global app_secret

    now = int(time.time())
    expire = now + expire_delta

    payload = "/increment:%d:%d" % (by, expire)
    payload_bytes = payload.encode('ascii')

    # Previously had used a base64 encoding...
    #hash_bytes = hmac.new(secret, payload_bytes, hashlib.sha256).digest()
    #hash_base64 = base64.b64encode(hash_bytes).decode('ascii')
    #print("hash = " + str(hash_base64))

    hash = hmac.new(app_secret, payload_bytes, hashlib.sha256).hexdigest()

    # FIXME: should properly urlencode params
    return "%s/increment?by=%d&expiry=%d&mac=%s" % (host, by, expire, hash)


if args.subcommand == "read":
    url = host + "/"
    r = requests.get(url)
    if r.status_code != 200:
        print("Error: " + r.text)
    else:
        print("Count = " + r.text)
elif args.subcommand == "increment":
    r = None
    if args.server_api:
        url = host + "/server_increment"
        r = requests.get(url, params = {
            "secret": server_secret,
            "by": args.by
        })
    else:
        url = create_signed_increment_url(args.by)
        r = requests.get(url)
    print("Status = " + str(r.status_code))
    if r.status_code == 200:
        print("New Count = " + r.text)
    else:
        print("Error: " + r.text)

elif args.subcommand == "reset":
    url = host + "/reset"
    r = requests.get(url, params = {
        "secret": server_secret
    })
    print("Status = " + str(r.status_code))
    if r.status_code != 200:
        print(r.text)

elif args.subcommand == "daily_snapshots":
    url = host + "/daily_snapshots"
    print("url = " + url)
    r = requests.get(url)
    print("Status = " + str(r.status_code))
    if r.status_code == 200:
        print("Reply = " + json.dumps(r.json(), indent=4))
    else:
        print("Error = " + r.text)

elif args.subcommand == "hourly_snapshots":
    url = host + "/hourly_snapshots"
    r = requests.get(url)
    print("Status = " + str(r.status_code))
    if r.status_code == 200:
        print("Reply = " + json.dumps(r.json(), indent=4))
    else:
        print("Error = " + r.text)

elif args.subcommand == "test":
    # Smoke test the /increment endpoint...

    url = create_signed_increment_url(500)
    print("URL = " + url)

    r = requests.get(url)
    print("status = " + str(r.status_code) + ", text = " + r.text)

    # try replay...
    r = requests.get(url)
    print("replay attempt status = " + str(r.status_code) + ", text = " + r.text)
    assert(r.status_code != 200)

    # Try sending large increment...
    url = create_signed_increment_url(5000, expire_delta=0)
    r = requests.get(url)
    print("large request status = " + str(r.status_code) + ", text = " + r.text)
    assert(r.status_code != 200)

    # Try sending an expired request...
    url = create_signed_increment_url(500, expire_delta=0)
    time.sleep(10)
    r = requests.get(url)
    print("expired request status = " + str(r.status_code) + ", text = " + r.text)
    assert(r.status_code != 200)

test-signed-increment.py
#!/usr/bin/env python3

import sys
import time
import hmac
import hashlib
import base64
import requests
import urllib.parse
import json

config = None
with open('config.json') as fp:
    config = json.load(fp)
if config == None:
    sys.exit("Failed to load config.json")

host = config['host']
secret = config['increment_secret'].encode('UTF-8')


def create_signed_increment_url(by, expire_delta=60):
    global host
    global secret

    now = int(time.time())
    expire = now + expire_delta

    payload = "/increment:%d:%d" % (by, expire)
    #print("signed payload = " + payload)
    payload_bytes = payload.encode('ascii')

    # Previously had used a base64 encoding...
    #hash_bytes = hmac.new(secret, payload_bytes, hashlib.sha256).digest()
    #hash_base64 = base64.b64encode(hash_bytes).decode('ascii')
    #print("hash = " + str(hash_base64))

    hash = hmac.new(secret, payload_bytes, hashlib.sha256).hexdigest()
    #print("hash = " + str(hash))

    # FIXME: should properly urlencode params
    return "%s/increment?by=%d&expiry=%d&mac=%s" % (host, by, expire, hash)


# Smoke test the /increment endpoint...

url = create_signed_increment_url(500)
print("URL = " + url)

r = requests.get(url)
print("status = " + str(r.status_code) + ", text = " + r.text)

# try replay...
r = requests.get(url)
print("replay attempt status = " + str(r.status_code) + ", text = " + r.text)
assert(r.status_code != 200)

# Try sending large increment...
url = create_signed_increment_url(5000, expire_delta=0)
r = requests.get(url)
print("large request status = " + str(r.status_code) + ", text = " + r.text)
assert(r.status_code != 200)

# Try sending an expired request...
url = create_signed_increment_url(500, expire_delta=0)
time.sleep(10)
r = requests.get(url)
print("expired request status = " + str(r.status_code) + ", text = " + r.text)
assert(r.status_code != 200)

src/counter.mjs

/*
function byteStringToUint8Array(byteString) {
  const ui = new Uint8Array(byteString.length)
  for (let i = 0; i < byteString.length; ++i) {
    ui[i] = byteString.charCodeAt(i)
  }
  return ui
}
*/

function hexDigestToUintArray(hexDigest) {
  const nBytes = hexDigest.length / 2;
  const ui = new Uint8Array(nBytes);
  for (let i = 0; i < nBytes; i++) {
    ui[i] = parseInt(hexDigest.substring(i * 2, (i * 2) + 2), 16);
  }
  return ui
}

function compareArrays(a, b) {
  if (a === null || b === null)
    return false;

  if (a.length !== b.length)
    return false;

  const length = a.length;
  for (let i = 0; i < length; i++) {
    if (a[i] != b[i])
      return false;
  }

  return true;
}

function float2int (value) {
  return value | 0;
}

export class Counter {
  constructor(state, env) {

    this.encoder = new TextEncoder();
    this.appSecret = this.encoder.encode(env.APP_SECRET);
    this.serverSecret = env.SERVER_SECRET;

    this.maxIncrement = 2000;

    this.historyPos = 0;
    this.history = new Array(10).fill(null);

    this.maxHourlySnapshots = 24;
    this.maxDailySnapshots = 30;

    this.state = state;
  }


  async do_increment(request, by) {
      let state = await this.state.storage.get(["value", "daily_snapshots", "daily_snapshots_count", "hourly_snapshots", "hourly_snapshots_count"]);


      let currentCount = state.get("value") || 0;
      let dailySnapshots = state.get("daily_snapshots") || [];
      let dailySnapshotsCount = state.get("daily_snapshots_count") || 0; // New snapshots count (goes beyond max length for snapshots!)
      let hourlySnapshots = state.get("hourly_snapshots") || [];
      let hourlySnapshotsCount = state.get("hourly_snapshots_count") || 0;

      // index snapshots with an ephoch timestamps in seconds...
      let now = new Date();
      let dailySnapshotTime = float2int(new Date(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDay()).valueOf() / 1000);
      let hourlySnapshotTime = float2int(new Date(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDay(), now.getUTCHours()).valueOf() / 1000)

      if (dailySnapshotsCount <= 0) {
        dailySnapshots.push([dailySnapshotTime, 0])
        dailySnapshotsCount = 1;
      }
      if (hourlySnapshotsCount <= 0) {
        hourlySnapshots.push([hourlySnapshotTime, 0])
        hourlySnapshotsCount = 1;
      }

      let dailyPos = (dailySnapshotsCount - 1) % this.maxDailySnapshots;
      let hourlyPos = (hourlySnapshotsCount - 1) % this.maxHourlySnapshots;

      let [latestDailyTimestamp, latestDailyCount] = dailySnapshots[dailyPos];
      let [latestHourlyTimestamp, latestHourlyCount] = hourlySnapshots[hourlyPos];


      // Either append the count if we're still in the same day, or create a new snapshot...
      if (latestDailyTimestamp == dailySnapshotTime) {
        dailySnapshots[dailyPos] = [latestDailyTimestamp, latestDailyCount + by];
      } else {
        if (dailySnapshots.length < this.maxDailySnapshots) {
          dailySnapshots.push([dailySnapshotTime, by])
        } else {
          dailyPos += 1;
          dailyPos %= this.maxDailySnapshots;
          dailySnapshots[dailyPos] = [dailySnapshotTime, latestDailyCount + by];
        }
        dailySnapshotsCount += 1
      }

      // Either append the count if we're still in the same hour, or create a new snapshot...
      if (latestHourlyTimestamp == hourlySnapshotTime) {
        hourlySnapshots[hourlyPos] = [latestHourlyTimestamp, latestHourlyCount + by];
      } else {
        if (hourlySnapshots.length < this.maxHourlySnapshots) {
          hourlySnapshots.push([hourlySnapshotTime, by])
        } else {
          hourlyPos += 1;
          hourlyPos %= this.maxHourlySnapshots;
          hourlySnapshots[hourlyPos] = [hourlySnapshotTime, latestHourlyCount + by];
        }
        hourlySnapshotsCount += 1
      }

      this.state.storage.put("value", currentCount + by);
      this.state.storage.put("hourly_snapshots", hourlySnapshots);
      this.state.storage.put("hourly_snapshots_count", hourlySnapshotsCount);
      this.state.storage.put("daily_snapshots", dailySnapshots);
      await this.state.storage.put("daily_snapshots_count", dailySnapshotsCount);  // Only await here to get a write coallece / transaction ...

      return new Response(currentCount + by);
  }

  async handle_app_increment(request, url) {
    if (!url.searchParams.has("mac") || !url.searchParams.has("expiry") || !url.searchParams.has("by")) {
      return new Response("Missing query parameter", { status: 403 })
    }

    const key = await crypto.subtle.importKey(
      "raw",
      this.appSecret,
      { name: "HMAC", hash: "SHA-256" },
      false,
      ["verify"],
    )

    const expiry = parseInt(url.searchParams.get("expiry"));
    if (expiry === NaN || expiry <= 0) {
      const body = "Invalid Expiry"
      return new Response(body, { status: 400 })
    }

    const by = parseInt(url.searchParams.get("by"));
    if (by === NaN || by <= 0 || by > this.maxIncrement) {
      const body = "Invalid increment 'by' count";
      return new Response(body, { status: 400 })
    }

    // E.g. string being SHA-256 signed is like "/increment:500:12345678"
    // and then request is made like: https://api.foo.bar/increment?expiry=12345678&by=500&mac=f1a20f01bb
    const dataToAuthenticate = url.pathname + ":" + by + ":" + expiry;

    const receivedMac = hexDigestToUintArray(url.searchParams.get("mac"));

    // Initially I passed the hmac via a base64 string but figured it should be less
    // error prone for clients to send a hex digest (e.g. less chance of bugs with
    // url encoding '+' characters)
    //
    //const receivedMacBase64 = url.searchParams.get("mac")
    //const receivedMac = byteStringToUint8Array(atob(receivedMacBase64))
    //let receivedMac = null;
    //try {
    //  receivedMac = Uint8Array.from(atob(receivedMacBase64), (v) => v.charCodeAt(0));
    //} catch (e) {
    //  const body = "Invalid MAC"
    //  return new Response(body, { status: 403 })
    //}

    // Check for simple replay attempts...
    for (let i = 0; i < this.history.length; i++) {
      if (compareArrays(this.history[i], receivedMac)) {
        const body = "Ignoring duplicate request";
        return new Response(body, { status: 400 })
      }
    }

    const verified = await crypto.subtle.verify(
      "HMAC",
      key,
      receivedMac,
      this.encoder.encode(dataToAuthenticate),
    )

    if (!verified) {
      const body = "Invalid MAC"
      return new Response(body, { status: 403 })
    }

    if (Date.now() > (expiry * 1000)) {
      const body = `URL expired at ${new Date(expiry * 1000)}`
      return new Response(body, { status: 403 })
    }

    let resp = await this.do_increment(request, by)

    // NB: cloudflare will block the event loop for this worker until
    // this returns (so no other fetch requests will be seen here
    // for this durable object until this is done
    //
    // (input gate rules)

    // Simple circular buffer of recent hashes we can check to thwart simple
    // replay attempts...
    // Note: we don't worry about writing this to storage, an ephemeral in-memory history
    // should be fine here...
    // Note: we only append to the history after the storage has completed
    this.history[this.historyPos++] = receivedMac;
    this.historyPos %= this.history.length;

    return resp;
  }

  async handle_server_increment(request, url) {
    if (!url.searchParams.has("secret") || !url.searchParams.has("by")) {
      return new Response("Missing query parameter", { status: 403 })
    }

    const secret = url.searchParams.get("secret");
    if (secret !== this.serverSecret)
    {
      return new Response("Invalid secret", { status: 403 })
    }

    const by = parseInt(url.searchParams.get("by"));
    if (by === NaN || by <= 0 || by > this.maxIncrement) {
      const body = "Invalid increment 'by' count";
      return new Response(body, { status: 400 })
    }

    return await this.do_increment(request, by)
  }

  async handle_get_snapshots(request, url, name) {
    let state = await this.state.storage.get([name + "_snapshots", name + "_snapshots_count"]);
    let snapshotsBuffer = state.get(name + "_snapshots") || [];
    let snapshotsCount = state.get(name + "_snapshots_count") || 0;
    let start = snapshotsCount % snapshotsBuffer.length;

    // Reorder, instead of sending the circular buffer directly...
    let snapshots = new Array(snapshotsBuffer.length);
    for (let i = 0; i < snapshotsBuffer.length; i++) {
      let pos = (start + i) % snapshotsBuffer.length;
      snapshots[i] = snapshotsBuffer[pos];
    }
    let data = {
      ver: 1,
      snapshots: snapshots
    }

    const json = JSON.stringify(data, null, 2)
    return new Response(json, {
      headers: {
        "content-type": "application/json;charset=UTF-8"
      }
    })
  }

  // Handle HTTP requests from clients.
  async fetch(request) {

    let url = new URL(request.url);

    switch (url.pathname) {
    case "/increment":
      return await this.handle_app_increment(request, url);

    // As an alternative to the /increment endpoint that is used directly by clients/apps where
    // there is more chance of users being able to see those URLs and attempt to forge increment
    // requests, this API can be used by backends simply based on a shared secret that's never
    // exposed to apps
    case "/server_increment":
      return await this.handle_server_increment(request, url);

    case "/reset":

      if (!url.searchParams.has("secret")) {
        return new Response("Missing query parameter", { status: 403 })
      }

      let secret = url.searchParams.get("secret");
      if (secret !== this.serverSecret)
      {
        return new Response("Invalid secret", { status: 403 })
      }

      this.state.storage.put("value", 0);
      this.state.storage.put("hourly_snapshots", []);
      this.state.storage.put("hourly_snapshots_count", 0);
      this.state.storage.put("daily_snapshots", []);
      await this.state.storage.put("daily_snapshots_count", 0);  // Only await here to get a write coallece / transaction ...

      return new Response(0);

    case "/":
      var currentCount = await this.state.storage.get("value") || 0;
      return new Response(currentCount);

    case "/daily_snapshots":
      return await this.handle_get_snapshots(request, url, "daily")

    case "/hourly_snapshots":
      return await this.handle_get_snapshots(request, url, "hourly")


    default:
      return new Response("Not found", {status: 404});
    }

    return new Response("Bad Request", {status: 400});
  }
}
